pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,126 @@
+package tempdir
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/storage/pkg/lockfile"
+	"github.com/sirupsen/logrus"
+)
+
+// TODO: implement alternative for: system.EnsureRemoveAll(dir)
+
+// TempDir represents a temporary directory that is created in a specified root directory.
+type TempDir struct {
+	rootDir      string
+	tempDirPath  string
+	lockFilePath string
+	lock         *lockfile.LockFile
+}
+
+// recover checks for stale temporary directories in the specified root directory.
+// If any stale directories are found, they are removed.
+// Stale directories are identified by their naming pattern "temp-dir-*",
+// and the function attempts to acquire a lock on each directory before
+// removing it. If a lock cannot be acquired, the directory is skipped, because other
+// processes may still be using it.
+//
+// It is important to note that this function should be called before
+// creating a new temporary directory to ensure a clean state.
+//
+// Required global storage lock must be held before calling this function.
+func recover(rootDir string) error {
+	tempDirPattern := filepath.Join(rootDir, "temp-dir-*")
+	tempDirs, err := filepath.Glob(tempDirPattern)
+	if err != nil {
+		return fmt.Errorf("error finding tmp dirs in %s: %w", rootDir, err)
+	}
+
+	var recoveryErrors []error
+
+	for _, tempDirPath := range tempDirs {
+		lockFilePath := filepath.Join(tempDirPath, ".lock")
+		lock, err := lockfile.GetLockFile(lockFilePath)
+		if err != nil {
+			recoveryErrors = append(recoveryErrors, fmt.Errorf("error getting lock file %s: %w", lockFilePath, err))
+			continue
+		}
+
+		if err := lock.TryLock(); err == nil {
+			lock.Unlock()
```

Comment:
Why the unlock _here_? That way a concurrent `recover` starting at this point will _also_ try to remove things, that seems unnecessary.

“The process that holds the lock is responsible for deleting the related subdirectory” might be a good invariant to hold, but I’m not sure what design you are working towards.